llvm/lib/CodeGen/LiveDebugValues/InstrRefBasedImpl.cpp

```diff
@@ -3503,7 +3504,10 @@ bool InstrRefBasedLDV::depthFirstVLocAndEmit(
   // Helper lambda for ejecting a block -- if nothing is going to use the block,
   // we can translate the variable location information into DBG_VALUEs and then
   // free all of InstrRefBasedLDV's data structures.
+  SmallPtrSet<const MachineBasicBlock *, 8> EjectedBBs;
   auto EjectBlock = [&](MachineBasicBlock &MBB) -> void {
+    if (EjectedBBs.insert(&MBB).second == false)
+      return;
```

Comment:
My recollection of this (which may be faulty) is that blocks should only get ejected once, in the order determined just above this block (it reduces the working-set size during compilation). If the added code fires I'd suggest there's a bug somewhere else, is there a reproducer for that? I'd imagine it'd only reduce performance a bit.

Reply:
The rationale for this code is because I removed the code [in line 3596](https://github.com/llvm/llvm-project/pull/74203/files#diff-c952216f21020db4d436ad0aa579a75aee29aab3ac1fe07a6541246e2bb38395R3596) of the "green" side of the diff.

I.e., before:
```
 for (auto *MBB : ArtificialBlocks)
     if (MOutLocs[MBB->getNumber()])
       EjectBlock(*MBB);
```

After:
```
  for (auto *MBB : ArtificialBlocks)
     EjectBlock(*MBB);
```

Prior to this change, we were testing whether something had been ejected by querying whether we had called "reset" on the unique pointer. There are no unique pointers anymore, and checking for whether the length of the vector is 0 not the equivalent, so the only solution was to add this check.

So, with this mind, the patch is 100% intended to be NFC. Do you agree with that?

Reply:
Ahhh, I see the rationale. I'm struggling to remember where that came from -- I thought it should be the case that the set of ArtificialBlocks and the set of blocks produced by `makeDepthFirstEjectionMap` shouldn't overlap anyway because the artificial blocks shouldn't be in any lexical scopes? But maybe they can, as LexicalScopes only stores a start/end block-number pair.

Either way, I'm not confident enough to say the is-null test isn't double-ejection prevention so it should be kept. For performance, could you use a `SmallDenseMap` instead of SmallPtrSet with a call to `reserve` as we'll know the maximum number of blocks that might get inserted. That limits the worst-case map-growth to a single allocation while providing a local scratchpad for small functions.

Reply:
> I'm not confident enough to say the is-null test isn't double-ejection prevention so it should be kept

I think there may be some miscommunication going here: the check _is kept_, it just takes a different form.
I moved it _inside_ the lambda, and instead of checking for a nullptr (pointers don't exist anymore, so we can't check that), I am using an "has been ejected" check.

Does that make sense?

Reply:
Makes perfect sense and you're correct; I'm just second-guessing whether the check had a different purpose, or was something I'd gratuitously added in the past without reason.